python/tk_maya/panel_generation.py

```diff
@@ -0,0 +1,145 @@
+        if pm.workspaceControlState(maya_panel_id, exists=True):
+            engine.log_debug("Deleting existing Maya workspace panel state %s." % maya_panel_id)
+            pm.workspaceControlState(maya_panel_id, remove=True)
+
+        # Retrieve the Channel Box dock area, with error reporting turned on.
+        # This MEL function is declared in Maya startup script file UIComponents.mel.
+        dock_area = mel.eval('getUIComponentDockControl("Channel Box / Layer Editor", true)')
+        engine.log_debug("Retrieved Maya dock area %s." % dock_area)
+
+        # This UI script will be called to build the UI of the new dock tab.
+        # It will embed the Shotgun app panel widget into a Maya workspace control.
+        # Maya 2017 expects this script to be passed in as a string, not as a function pointer
+        # according to C++ source code method TelfWorkspaceControlCmd::handleScriptFlag().
+        # See function _build_workspace_control_ui() below for a commented version of this script.
+        ui_script = "import pymel.core as pm\n" \
+                    "workspace_control = pm.setParent(query=True)\n" \
+                    "pm.control('%s', edit=True, parent=workspace_control)" \
+                    % widget_id
+
+        # The following UI script can be used for development and debugging purposes.
+        # This script has to retrieve and import the current source file in order to call
+        # function _build_workspace_control_ui() below to build the workspace control UI.
+        # ui_script = "import imp\n" \
+        #             "panel_generation = imp.load_source('%s', '%s')\n" \
+        #             "panel_generation._build_workspace_control_ui('%s')" \
+        #             % (__name__, __file__.replace(".pyc", ".py"), widget_id)
+
+        # Dock the Shotgun app panel widget into a new tab of the Channel Box dock area.
+        engine.log_debug("Creating Maya workspace panel %s." % maya_panel_id)
+        dock_tab = pm.workspaceControl(maya_panel_id,
+                                       tabToControl=(dock_area, -1),  # -1 to append a new tab
+                                       uiScript=ui_script,
+                                       loadImmediately=True,
+                                       retain=False,  # delete the dock tab when it is closed
+                                       label=title,
+                                       r=True  # raise the new dock tab to the top
+                   )
+
+        # Now that the workspace dock tab has been created, let's update its UI script.
+        # This updated script will be saved automatically with the workspace control state
+        # in the Maya layout preference file when the user will choose to quit Maya,
+        # and will be executed automatically when Maya is restarted later by the user.
+
+        # The script will delete the empty workspace dock tab that Maya will recreate on startup
```

Comment:
similar to my comment above, instead of deleting, let's discuss if we can preserve this state and fire up the toolkit windows correctly on workspace restore.